packages/sitecore-jss/src/personalization/layout-personalization-service.ts

```diff
@@ -0,0 +1,220 @@
+    context: PersonalizationContext,
+    route: RouteData
+  ): Promise<PersonalizationResult> {
+    // clear personalization before getting new one
+    this.personalizationResult = {};
+
+    const personalizedRenderings = this.layoutPersonalizationUtils.getPersonalizableComponents(
+      route.placeholders
+    );
+
+    if (!personalizedRenderings.length) {
+      return { isPersonalizable: false };
+    }
+
+    const currentResult = this.personalizationResult;
+    currentResult.personalizationOperation = this.personalizeComponents(
+      {
+        routePath: context.routePath,
+        language: context.language,
+      },
+      personalizedRenderings
+    );
+
+    try {
+      const components = await currentResult.personalizationOperation;
+      currentResult.components = components;
+      return { isPersonalizable: true };
+    } catch (error) {
+      currentResult.personalizationOperation = undefined;
+      throw error;
+    }
+  }
+
+  /**
+   * Gets the personalized component.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  getPersonalizedComponent(componentUid: string): ComponentRendering | null {
+    if (!this.personalizationResult.components) {
+      return null;
+    }
+
+    return this.personalizationResult.components[componentUid] ?? null;
+  }
+
+  /**
+   * Provides a value that indicates whether the loading is in-progress.
+   * @returns {boolean} The value that indicates whether the loading is in-progress.
+   */
+  isLoading(): boolean {
+    return (
+      !!this.personalizationResult.personalizationOperation &&
+      !this.personalizationResult.components
+    );
+  }
+
+  /**
+   * Ensures the personalized component is loaded.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  async ensurePersonalizedComponentLoaded(
+    componentUid: string
+  ): Promise<ComponentRendering | null> {
+    if (!this.personalizationResult.personalizationOperation) {
+      throw new Error(
+        `${this.fetchPersonalization.name} should be called before getting personalized component`
+      );
+    }
+
+    const personalizedComponents = await this.personalizationResult.personalizationOperation;
+    if (!personalizedComponents) {
+      return null;
+    }
+
+    return personalizedComponents[componentUid] ?? null;
+  }
+
+  /**
+   * Personalizes components.
+   * @param {PersonalizationContext} context The context.
+   * @param {PersonalizableComponentRendering[]} personalizableRenderings The personalizable components.
+   * @returns {Object.<string, ComponentRendering | null>} The personalized components.
+   */
+  async personalizeComponents(
+    context: PersonalizationContext,
+    personalizableRenderings: PersonalizableComponentRendering[]
+  ): Promise<{ [key: string]: ComponentRendering | null }> {
```

Comment:
This method is marked as returning a Promise, but it does not always return a Promise

Reply:
Async functions always return a promise. If the return value of an async function is not explicitly a promise, it will be implicitly wrapped in a promise.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function